src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala

```diff
@@ -318,14 +318,15 @@ trait SyntheticMethods extends ast.TreeDSL {
 
     /* If you serialize a singleton and then deserialize it twice,
      * you will have two instances of your singleton unless you implement
-     * readResolve.  Here it is implemented for all objects which have
-     * no implementation and which are marked serializable (which is true
-     * for all case objects.)
+     * readResolve.  Here we use a serialization proxy that prevents
+     * serialization of state and implements readResolve
+     * for all objects which have no implementation and which are marked
+     * serializable (which is true for all case objects.)
      */
-    def needsReadResolve = (
+    def needsModuleSerializationProxy = (
          clazz.isModuleClass
       && clazz.isSerializable
-      && !hasConcreteImpl(nme.readResolve)
+      && !hasConcreteImpl(nme.writeReplace)
```

Comment:
this is not the correct test.
We should be checking for a matching method, not a method with this name

It should not match `def writeReplace(x:Int) = ???`

Reply:
We would need to rework significantly to avoid do a semantic check. We did manage to do that for `Companion.apply` recently, but that was far from easy. I'd suggest we take that up as a separate change.